File: Misc/116_lc_208.java
package Misc;

/*
 * 208. Implement Trie (Prefix Tree)
 */

class Node {
    public Node[] node;
    private boolean endFlag; // by default false

    public Node() {
        this.node = new Node[26];
    }

    public boolean containsKey(char c) {
        return node[c-'a'] != null;
    }

    public void putKey(char c) {
        node[c-'a'] = new Node();
    }

    public Node getCharReferenceNode(char c) {
        return node[c-'a'];
    }

    public boolean getEndFlag() {
        return this.endFlag;
    }

    public void setEndFlag(boolean val) {
        this.endFlag = val;
    }
}

class Trie {
    private Node root;
    public Trie() {
        root = new Node();
    }
    
    public void insert(String word) {
        Node node = root; // getting a new node to iterate otherwise main root will be shifted  
        for (char ch: word.toCharArray()) {
            // if char already exists in root then we only need to shift reference
            if (!node.containsKey(ch)) {
                node.putKey(ch);
            }
            node = node.getCharReferenceNode(ch); // shifting reference node
        }
        node.setEndFlag(true);
    }
    
    public boolean search(String word) {
        Node node = root;
        for (char ch: word.toCharArray()) {
            if (!node.containsKey(ch)) {
                return false;
            }
            node = node.getCharReferenceNode(ch);
        }
        return node.getEndFlag();
    }
    
    public boolean startsWith(String prefix) {
        Node node = root;
        for (char ch: prefix.toCharArray()) {
            if (!node.containsKey(ch)) {
                return false;
            }
            node = node.getCharReferenceNode(ch);
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: Misc/116_lc_208.py
class TrieNode:
    def __init__(self):
        self.node = [None for _ in range(26)]
        self.endFlag = False

    def containsKey(self, c):
        return self.node[ord(c) - ord("a")] != None

    def putKey(self, c):
        self.node[ord(c) - ord("a")] = TrieNode()

    def getCharReferenceNode(self, c):
        return self.node[ord(c) - ord("a")]

    def getEndFlag(self):
        return self.endFlag

    def setEndFlag(self, val):
        self.endFlag = val


class Trie:
    def __init__(self):
        self.rootNode = TrieNode()

    def insert(self, word):
        node = self.rootNode
        for char in word:
            if not node.containsKey(char):
                node.putKey(char)
            node = node.getCharReferenceNode(char)
        node.setEndFlag(True)

    def search(self, word):
        node = self.rootNode
        for char in word:
            if not node.containsKey(char):
                return False
            node = node.getCharReferenceNode(char)
        return node.getEndFlag()

    def startsWith(self, prefix):
        node = self.rootNode
        for char in prefix:
            if not node.containsKey(char):
                return False
            node = node.getCharReferenceNode(char)
        return True
